scientific/edf.ksy

```diff
@@ -0,0 +1,182 @@
+      ndr:
+        value: num_data_records.to_i
+
+    seq:
+      - id: "version"
+        size: 8
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "patient_id"
+        size: 80
+        type: str
+        encoding: ASCII
+
+      - id: "recording_id"
+        size: 80
+        type: str
+        encoding: ASCII
+
+      - id: "start_date"
+        size: 8
+        type: str
+        encoding: ASCII
+
+      - id: "start_time"
+        size: 8
+        type: str
+        encoding: ASCII
+
+      - id: "header_num_bytes"
+        size: 8
+        type: str
+        encoding: ASCII
+        pad-right: 0x20
+
+      - id: "reserved"
+        size: 44
+        type: str
+        encoding: ASCII
+
+      - id: "num_data_records"
+        size: 8
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "data_duration"
+        size: 8
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "num_signals"
+        size: 4
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "labels"
+        size: 16
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "transducers"
+        size: 80
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "phys_dimensions"
+        size: 8
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "phys_mins"
+        size: 8
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "phys_maxes"
+        size: 8
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "dig_mins"
+        size: 8
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "dig_maxes"
+        size: 8
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "prefiltering"
+        size: 80
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "num_samples"
+        size: 8
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+      - id: "signals_reserved"
+        size: 32
+        repeat: expr
+        repeat-expr: ns
+        type: str
+        pad-right: 0x20
+        encoding: ASCII
+
+  body:
+    types:
+      data_recording:
+        seq:
+          - id: "data"
+            type: u2
+            repeat: expr
+            repeat-expr: _root.header.num_samples[_parent.signal_index].to_i
+
+      signal:
+        params:
+          - id: "signal_index"
```

Comment:
1. Why not just `idx`?
2. Should it really be signed?

Reply:
Thanks for the info on how it is going on in JVM-based langs.

Reply:
> It is semantically incorrect to have negative indexes in an array.

Yes, but unsigned types are not a good solution to this.

Reply:
Unsigned types are only good solution for it and only they must be allowed there, any signed type used as an array index must be a syntax error. In the case of sunbtraction - if a compiler can prove that the result is non-negative, it should be cast to unsigned automatically. If the compiler cannot - it must be an error and the programmer must code it so that the compiler can prove it.

It is JVM and Java that are flawed, not unsigned numbers. Wasting 1 bit for a sign that must always be 0 used is waste.